File: sdks/typescript/tests/crypto-roundtrip.test.ts
import { describe, expect, it } from "vitest";
import {
  decryptForRecipient,
  encryptForRecipients,
  generateEd25519Keypair,
  generateX25519Keypair,
  signProtectedPayload,
  verifyProtectedPayloadSignature
} from "../src/crypto.js";
import { buildEnvelope } from "../src/messages.js";

describe("Crypto roundtrip", () => {
  it("encrypts, signs, verifies, and decrypts payload", () => {
    const senderSigning = generateEd25519Keypair();
    const recipientEncryption = generateX25519Keypair();
    const envelope = buildEnvelope({
      sender: "agent:sender@demo",
      recipients: ["agent:recipient@demo"],
      message_class: "SEND",
      context_id: "ctx:test",
      expires_in_seconds: 120,
      operation_id: "op:test"
    });
    const payload = {
      kind: "demo",
      sequence: 1
    };
    const encrypted = encryptForRecipients(payload, envelope, {
      "agent:recipient@demo": recipientEncryption.public_key
    });
    const signed = signProtectedPayload(envelope, encrypted, senderSigning.private_key, "sig:test");
    expect(verifyProtectedPayloadSignature(envelope, signed, senderSigning.public_key)).toBe(true);
    const decrypted = decryptForRecipient(
      envelope,
      signed,
      "agent:recipient@demo",
      recipientEncryption.private_key
    );
    expect(decrypted.kind).toBe("demo");
    expect(decrypted.sequence).toBe(1);
  });
});

File: sdks/go/crypto_roundtrip_test.go
package acp

import "testing"

func TestEncryptSignVerifyAndDecryptRoundtrip(t *testing.T) {
	senderSigningPrivate, senderSigningPublic, err := GenerateEd25519Keypair()
	if err != nil {
		t.Fatalf("unable to generate sender signing keys: %v", err)
	}
	recipientEncryptionPrivate, recipientEncryptionPublic, err := GenerateX25519Keypair()
	if err != nil {
		t.Fatalf("unable to generate recipient encryption keys: %v", err)
	}
	envelope, err := BuildEnvelope(EnvelopeInput{
		Sender:           "agent:sender@demo",
		Recipients:       []string{"agent:recipient@demo"},
		MessageClass:     MessageSend,
		ContextID:        "ctx:test",
		ExpiresInSeconds: 300,
		OperationID:      "op:roundtrip",
		Tenant:           "tenant.demo",
	})
	if err != nil {
		t.Fatalf("envelope should be created: %v", err)
	}
	if envelope.Tenant == nil || *envelope.Tenant != "tenant.demo" {
		t.Fatalf("envelope tenant should be present")
	}
	payload := map[string]any{
		"type":     "demo",
		"sequence": float64(1),
	}
	protectedPayload, err := EncryptForRecipients(payload, envelope, map[string]string{
		"agent:recipient@demo": recipientEncryptionPublic,
	})
	if err != nil {
		t.Fatalf("payload encryption should succeed: %v", err)
	}
	protectedPayload, err = SignProtectedPayload(envelope, protectedPayload, senderSigningPrivate, "sig:test")
	if err != nil {
		t.Fatalf("payload signature should succeed: %v", err)
	}
	if !VerifyProtectedPayloadSignature(envelope, protectedPayload, senderSigningPublic) {
		t.Fatalf("signature verification should pass")
	}
	decrypted, err := DecryptForRecipient(envelope, protectedPayload, "agent:recipient@demo", recipientEncryptionPrivate)
	if err != nil {
		t.Fatalf("recipient decryption should succeed: %v", err)
	}
	if decrypted["type"] != "demo" {
		t.Fatalf("decrypted payload should preserve type field")
	}
	if int(asFloat64FromFixture(t, decrypted["sequence"], "sequence must be numeric")) != 1 {
		t.Fatalf("decrypted payload should preserve numeric field")
	}
}
